src/overview.md

```diff
@@ -0,0 +1,368 @@
+(Note that the name is mostly historic. This is _not_ a "typing context" in the
+sense of `Γ` or `Δ` from type theory. The name is retained because that's what
+the name of the struct is in the source code.) All
+queries are defined as methods on the [`TyCtxt`] type, and the in-memory query
+cache is stored there too. In the code, there is usually a variable called
+`tcx` which is a handle on the typing context. You will also see lifetimes with
+the name `'tcx`, which means that something is tied to the lifetime of the
+`TyCtxt` (usually it is stored or _interned_ there).
+
+[`TyCtxt`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html
+
+### `ty::Ty`
+
+Types are really important in Rust, and they form the core of a lot of compiler
+analyses. The main type (in the compiler) that represents types (in the user's
+program) is [`rustc::ty::Ty`][ty]. This is so important that we have a whole chapter
+on [`ty::Ty`][ty], but for now, we just want to mention that it exists and is the way
+`rustc` represents types!
+
+Oh, and also the `rustc::ty` module defines the `TyCtxt` struct we mentioned before.
+
+[ty]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/type.Ty.html
+
+### Parallelism
+
+Compiler performance is a problem that we would like to improve on
+(and are always working on). One aspect of that is parallelizing
+`rustc` itself.
+
+Currently, there is only one part of rustc that is already parallel: codegen.
+During monomorphization, the compiler will split up all the code to be
+generated into smaller chunks called _codegen units_. These are then generated
+by independent instances of LLVM. Since they are independent, we can run them
+in parallel. At the end, the linker is run to combine all the codegen units
+together into one binary.
+
+However, the rest of the compiler is still not yet parallel. There have been
+lots of efforts spent on this, but it is generally a hard problem. The current
+approach is (**TODO: verify**) to turn `RefCell`s into `Mutex`s -- that is, we
+switch to thread-safe internal mutability. However, there are ongoing
+challenges with lock contention, maintaining query-system invariants under
+concurrency, and the complexity of the code base. One can try out the current
+work by enabling parallel compilation in `config.toml`. It's still early days,
+but there are already some promising performance improvements.
+
+### Bootstrapping
+
+`rustc` itself is written in Rust. So how do we compile the compiler? We use an
+older compiler to compile the newer compiler. This is called _bootstrapping_.
+
+Bootstrapping has a lot of interesting implications. For example, it means that one
+of the major users of Rust is Rust, so we are constantly testing our own
+software ("eating our own dogfood"). Also, it means building the compiler can
+take a long time because one must first build the compiler and then use that to
+build the new compiler (sometimes you can get away without the full 2-stage
```

Comment:
I guess people may be confused if they've never thought about this on the why this needs to be built twice. What if without a lot of detail we explain that in the first build you're compiling your program and second build is like you're executing your program.

Reply:
The explanation in the other chapter makes sense, but it's a bit long. Basically, the point is that we want to end up with a new compiler built with the new compiler. So the first stage is to compile the new compiler with the old compiler, and then we use stage 1 to build the new compiler again.

Reply:
@mark-i-m yeah, I know but the thing is how to briefly explain this without saying things that readers may not comprehend, hmmm. I kind of like the idea of stage 1 compile your "program", stage 2 execute your "program", the thing is that your "program" is a compiler which is what entangles all this. :/

Reply:
@spastorino Hmm... I guess, I don't quite see what you are getting at. stage 2 is also compiling no?

Reply:
added to followup issue